File: aa/ide.lua
-- hex
function hex_encode(txt)
	local res = {}
	for i=1,#txt do
		table.insert(res, string.format("%02x", string.byte(txt:sub(i,i))))
	end
	return table.concat(res)
end

function hex_decode(hex)
	local res = {}
	for i=1,#hex-1,2 do
		local sub = hex:sub(i,i+1)
		local num = tonumber(sub, 16)
		table.insert(res, string.char(num))
	end
	return table.concat(res)
end

-- file library
function open(path)
	local file = {}
	file.path = path
	file.cur = 1
	-- write => table, read => string
	file.data = nil
	
	function file:read(n)
		if not file.data then
			file.data = read(self.path) or read('/'..self.path) or error('could not read')
		end
		if file.cur >= #file.data then
			return nil
		end
		local res = self.data:sub(self.cur, self.cur + n - 1)
		if not res or #res == 0 then
			return nil
		end
		self.cur = self.cur + n
		return res
	end
	
	function file:write(res)
		if not file.data then
			file.data = {}
		end
		table.insert(file.data, res)
	end
	
	function file:close()
		if type(file.data) == 'table' then
			write(file.path, table.concat(file.data))
		end
	end
	
	return file
end

local term = js.global.document:getElementById('out')
function print(...)
	local args = {...}
	for i=1,#args do
		args[i] = tostring(args[i])
	end
	local text = table.concat(args, '\t')
	term.innerHTML = term.innerHTML .. text .. '<br>'
	term.scrollTop = 99999
end

function read(name)
	return js.global:read('/'..name)
end

function write(name, data)
	js.global:write('/'..name, data)
	js.global:fill(js.global.document:getElementById('tree'))
end

local reg = debug.getregistry()
function require(name)
	if reg._LOADED[name] then
		return
	end
	reg._LOADED[name] = true
	local path = name..'.lua'
	local file = read(path)
	if file then
		local chunk,err = load(file)
		if not chunk then
			reg._LOADED[name] = 'error'
			print(path..' '..err)
		else
			reg._LOADED[name] = chunk
			return chunk()
		end
	else
		return req(name)
	end
end

File: aa/ide.js
var meat = document.getElementById('meat');
var out = document.getElementById('out');

editor = CodeMirror.fromTextArea(meat, {
	lineNumbers: true,
	//gutters: ['breakpoints'],
	indentUnit: 4,
	mode: 'lua',
	lineWrapping: true,
	cursorScrollMargin: 10,
	tabSize: 4,
	indentWithTabs: true,
	theme: 'slang',
	styleActiveLine: true,
	extraKeys: {
		'F1': function() {run(); return false;},
		'Ctrl-S': function() {download(); return false;},
	}
});

//setInterval(run, 1000);

var names = document.getElementsByClassName('name');
for (var i = 0; i < names.length; i++) {
	names[i].addEventListener('click', (function(j) {
		return function (e) {
			file = names[j].innerHTML;
			load(file);
		};
	})(i));
}

function download() {
	var zip = new JSZip();
	addZip(zip, '/');
	var blob = zip.generate({type:"blob"});
	saveAs(blob, "project.zip");
}

function run() {
	read('/main.lua', function(code) {
	code = code || editor.getValue();
	try {
		L.execute(`
		debug.getregistry()._LOADED = {}
		js.global.document:getElementById('out').innerHTML = '';
		`)
		L.execute(code);
	} catch (e) {
		addLuaError(editor, e.lua_stack);
		out.innerHTML += e.lua_stack || e.toString();
		out.innerHTML += '\n';
	}
	});
}

// errors
var backgrounds = [];
var bookmarks = [];

function addError(editor,line,err) {
	// background
	var h = editor.addLineClass(line, 'background', 'lua-error');
	backgrounds.push(h);
	
	// message
	var e = document.createElement('span');
	e.className = 'lua-error-message';
	if (err.length > 20)
		err = err.substr(0,20)+'...';
	e.innerHTML = err;
	var bm = editor.setBookmark({line: line, pos: 0}, {
		widget: e,
		insertLeft: true,
	});
	bookmarks.push(bm);
}

function removeErrors(editor) {
	// clear existing errors
	for (var i = 0; i < backgrounds.length; i++)
		editor.removeLineClass(backgrounds[i], 'background', 'lua-error');
	for (var i = 0; i < bookmarks.length; i++)
		bookmarks[i].clear();
	
	// reset
	backgrounds = [];
	bookmarks = [];
}

function addLuaError(editor, message) {
	// [string 'a']:1: undefined symbol
	var split = message.split(':');
	var line = parseInt(split[1]) - 1;
	//if (!split[2])
	//	alert(message);
	var msg = split[2].substr(1);
	addError(editor,line,msg);
}

function check() {
	removeErrors(editor);
	try {
		L.load(editor.getValue());
	} catch (e) {
		addLuaError(editor, e.message);
	}
}

function save(path) {
	var data = editor.getValue();
	write(path, data);
}

editor.on('change', function() {
	if (selected.path.endsWith('.lua'))
	{check(); }
	if (!selected.path.endsWith('.exe'))
		save(selected.path);
});

get('/ide/ide.lua', function(src) {
	L.execute(src);
});
